prov/gni/src/gnix_mr.c

```diff
@@ -771,127 +973,292 @@ int _gnix_mr_cache_flush(gnix_mr_cache_t *cache)
+
+			rc = rbtErase(cache->stale.rb_tree, iter);
+			if (unlikely(rc != RBT_STATUS_OK)) {
+				GNIX_ERR(FI_LOG_MR,
+						"failed to erase entry from stale tree\n");
+			} else {
+				dlist_remove(&mr_entry->lru_entry);
+
+				atomic_dec(&cache->stale.elements);
+
+				__mr_cache_entry_destroy(mr_entry);
+			}
+
+			*entry = tmp;
 		} else {
-			atomic_inc(&cache->inuse_elements);
-			*entry = current_entry;
-			ret = FI_SUCCESS;
-		}
-	}
+			GNIX_INFO(FI_LOG_MR,
+					"removing entry from stale and migrating to inuse, "
+					"key=%llu:%llu\n",
+					mr_key->address, mr_key->length);
+			rc = rbtErase(cache->stale.rb_tree, iter);
+			if (unlikely(rc != RBT_STATUS_OK))
+				GNIX_WARN(FI_LOG_MR,
+						"failed to erase entry from stale tree\n");
+			assert(rc == RBT_STATUS_OK);
 
-	return ret;
-}
+			dlist_remove(&mr_entry->lru_entry);
 
-static int __mr_cache_lookup_inuse_slowpath(
-		gnix_mr_cache_t *cache,
-		gnix_mr_cache_key_t *key,
-		gnix_mr_cache_entry_t **entry)
-{
-	int ret = -FI_ENOENT;
+			atomic_dec(&cache->stale.elements);
 
-	return ret;
-}
+			/* if we made it to this point, there weren't
+			 * any entries in the inuse tree that would
+			 * have overlapped with this entry
+			 */
+			rc = rbtInsert(cache->inuse.rb_tree,
+					&mr_entry->key, mr_entry);
+			if (unlikely(rc != RBT_STATUS_OK))
+				GNIX_WARN(FI_LOG_MR,
+						"failed to insert entry into inuse tree\n");
+			assert(rc == RBT_STATUS_OK);
 
-static int __mr_cache_lookup_stale_slowpath(
-		gnix_mr_cache_t *cache,
-		gnix_mr_cache_key_t *key,
-		gnix_mr_cache_entry_t **entry)
-{
-	int ret = -FI_ENOENT;
+			atomic_set(&mr_entry->ref_cnt, 1);
+			atomic_inc(&cache->inuse.elements);
 
-	return ret;
+			*entry = mr_entry;
+		}
+
+		return FI_SUCCESS;
+	}
+
+	GNIX_INFO(FI_LOG_MR,
+			"could not use matching entry, "
+			"found=%llu:%llu\n",
+			mr_key->address, mr_key->length);
+
+	return -FI_ENOENT;
 }
 
 static int __mr_cache_create_registration(
 		gnix_mr_cache_t          *cache,
-		struct gnix_fid_mem_desc *mr,
 		struct gnix_fid_domain   *domain,
 		uint64_t                 address,
 		uint64_t                 length,
 		gni_cq_handle_t          dst_cq_hndl,
 		uint32_t                 flags,
 		uint32_t                 vmdh_index,
-		gni_mem_handle_t         *mem_hndl,
-		gnix_mr_cache_entry_t    **entry)
+		gnix_mr_cache_entry_t    **entry,
+		gnix_mr_cache_key_t      *key)
 {
 	int rc;
 	struct gnix_nic *nic;
 	gni_return_t grc = GNI_RC_SUCCESS;
-	gnix_mr_cache_entry_t    *current_entry;
+	gnix_mr_cache_entry_t *current_entry;
 
 	/* if we made it here, we didn't find the entry at all */
 	current_entry = calloc(1, sizeof(*current_entry));
 	if (!current_entry)
 		return -FI_ENOMEM;
 
+	dlist_init(&current_entry->lru_entry);
+	dlist_init(&current_entry->children);
+	dlist_init(&current_entry->siblings);
+
+	/* TODO: should we just try the first nic we find? */
 	/* NOTE: Can we assume the list is safe for access without a lock? */
```

Comment:
The domain should hold a reference to each NIC.  I think the MR cache needs to hold a reference to the domain so it is not destroyed and it's NIC list torn down.  If that's all true, do you want to remove this comment?